libraries/chain/fork_database.cpp

```diff
@@ -602,9 +558,80 @@ namespace eosio::chain {
       return bsp();
    }
 
+   // ------------------ fork_database -------------------------
+
+   fork_database::fork_database(const std::filesystem::path& data_dir)
+      : data_dir(data_dir) {
+   }
+
+   fork_database::~fork_database() {
+      apply<void>([&](auto& forkdb) { forkdb.close(data_dir / config::forkdb_filename); });
+   }
+
+   void fork_database::open( validator_t& validator ) {
+      std::lock_guard g(m);
+      if (!std::filesystem::is_directory(data_dir))
+         std::filesystem::create_directories(data_dir);
+
+      auto fork_db_file = data_dir / config::forkdb_filename;
+      if( std::filesystem::exists( fork_db_file ) ) {
+         try {
+            fc::cfile f;
+            f.set_file_path(fork_db_file);
+            f.open("rb");
+
+            fc::cfile_datastream ds(f);
+
+            // determine file type, validate totem
+            uint32_t totem = 0;
+            fc::raw::unpack( ds, totem );
+            EOS_ASSERT( totem == fork_database_legacy_t::legacy_magic_number || totem == fork_database_if_t::magic_number, fork_database_exception,
+                        "Fork database file '${filename}' has unexpected magic number: ${actual_totem}. Expected ${t1} or ${t2}",
+                        ("filename", fork_db_file)
+                        ("actual_totem", totem)("t1", fork_database_legacy_t::legacy_magic_number)("t2", fork_database_if_t::magic_number)
+            );
+
+            if (totem == fork_database_legacy_t::legacy_magic_number) {
+               apply_dpos<void>([&](auto& forkdb) {
```

Comment:
No need to store the magic number?

Reply:
We are not creating a `fork_database_t` here so it maintains the `magic_number` it was created with which was `legacy_magic_number`. Would be nice to create `fork_database_t` here, however that is not possible without adding a `std::monostate` to the variant or making the variant optional.

This inspired me to use a `std::optional` which helped point out that the current impl methods `apply_if` and `apply_dpos` were not thread safe.

Reply:
:( optional doesn't work because we expect a `nullptr` `chain_head` before fork database is opened.